File: 803-cheapest-flights-within-k-stops/cheapest-flights-within-k-stops.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {

        int[][] costArray = new int[n][k + 2];
        for (int[] row : costArray) Arrays.fill(row, Integer.MAX_VALUE);
        

        List<List<int[]>> adj = new ArrayList<>();
        for(int i=0; i<n; i++) adj.add(new ArrayList<>());

        for(int[] flight : flights) {
            adj.get(flight[0]).add(new int[]{flight[1],flight[2]});
        }

        PriorityQueue<int[]> pq = new PriorityQueue<>((x,y) -> Integer.compare(x[1],y[1]));

        pq.add(new int[]{src,0,0});
        costArray[src][0] = 0;

        while(!pq.isEmpty()) {
            int[] curr = pq.poll();
            int city = curr[0], cost = curr[1], stops = curr[2];

            
            if(city == dst) return cost;
            if(stops > k) continue;
            
            for(int[] neigh: adj.get(city)) {
                int nextFlights  = stops + 1;
                if(cost+neigh[1] < costArray[neigh[0]][nextFlights]) {
                    pq.add(new int[]{neigh[0],cost+neigh[1],nextFlights});
                    costArray[neigh[0]][nextFlights] = cost+neigh[1];
                }  
            }
        }

      
        return -1;
     }
}

File: 803-cheapest-flights-within-k-stops/cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        adj = defaultdict(list)
        for u,v,price in flights:
            adj[u].append((v,price))
        
        dist = [float('inf')]*n
        dist[src] = 0
        que = deque([(src,0)])
        level = 0

        while que and level <= k:
            size = len(que)
            for _ in range(size):
                path ,price = que.popleft()
                for neighbor_path,neighbor_price in adj[path]:
                    if neighbor_price  + price < dist[neighbor_path]:
                        dist[neighbor_path] = neighbor_price  + price
                        que.append((neighbor_path,dist[neighbor_path]))
            level += 1 
        return dist[dst] if dist[dst] != float('inf') else -1


 